website/content/functional-database.md

```diff
@@ -0,0 +1,299 @@
+  input requested and no system calls made.
+
+The most obvious effect is that for a pure function `my_function`, this:
+
+```js
+res1 = my_function(arg);
+res2 = my_function(arg);
+res3 = my_function(arg);
+```
+
+... is equivalent to:
+
+```js
+res1 = my_function(arg);
+res2 = res1;
+res3 = res2;
+```
+
+Because `my_function` is guaranteed to return the same result when given `arg`,
+we can skip calling `my_function` for the second and the third time, because we
+(or even better: the compiler) already know what the output will be.
+
+This may seem like a minor point, but we just getting started.
+
+### First class citizens
+
+Nearly a buzz word, the saying "functions are first class citizens" means that
+they are treated the same as any other value.
+
+They can be stored in a variable, they can be passed to another function as an
+argument or even be used in binary operations.
+
+```js
+function double(x) {
+  return 2 * x;
+}
+
+let preprocess = double;
+
+my_array = [4, 5, 6];
+
+my_array.map(preprocess);
+// yields [8, 10, 12]
+```
+
+Here, we've stored function `double` in variable `preprocess` and then passed
+that to method `Array.map`, which applies a given function to each of the array
+elements.
+
+### Currying
+
+Named after Haskell Curry, currying is an implicit act of converting a function
+call with missing arguments into a new function.
+
+```js
+function add(x, y) {
+  return x + y;
+}
+let add_one = add(1);
+```
+
+Because we didn't specify `y` parameter of `add` function, the result is a new
+function that is still waiting for the last argument. It is equivalent to
+defining add_one like this:
+
+function add_one(y) { return add(1, y) }
+
+### Implicit function call
+
+Let's suppose that we have a pure function that doesn't take any arguments:
+
+```js
+function my_function() {
+  return 42;
+}
+```
+
+Because it is a pure function, it cannot depend on anything other than its
+non-existing arguments. Another way to phrase it, this is a constant expression.
+
+And as such, it might as well be defined as one:
+
+```js
+let my_var = 42;
+```
+
+In conventional programming languages[1], there is a difference between using
+`my_function()` and `my_var`. The first one evaluates the expression at the call
+site, while the second one evaluates it at the declaration site. It is also
+possible to express just `my_function`, a reference to a function that can be
+called without arguments.
+
+But if all your functions are pure, we generalize by saying that all three cases
+are equivalent.
+
+To make this work, let's say that `my_function` is "implicitly invoked" just as
+it would have been expressed as `my_function()`.
+
+Also, if we declare all functions to be pure it doesn't matter exactly _when_
+they are evaluated. So let's give the compiler the authority to make an informed
+guess about _when_ and make `my_var` and `my_function()` semantically
+equivalent.
+
+[1]: Let's say that conventional languages are first 15 from this list:
```

Comment:
```suggestion
[^1]: Let's say that conventional languages are first 15 from this list:
```

(unless I'm wrong about markdown and this isn't needed??)

Reply:
I'm pretty sure I wrote it as `[^1]`... Strange